Stock available immediately2

gfcapalbo opened the pull request:
Port to 8.0

lepistone commented on stock_available_immediately/product.py:
```diff
@@ -26,11 +26,9 @@
 
 class product_immediately_usable(orm.Model):
```
please rename the class to `Product` (or `ProductTemplate`)

lepistone commented on stock_available_immediately/product.py:
```diff
@@ -55,7 +53,7 @@ def _product_available(self, cr, uid, ids, field_names=None,
             for product_id, stock_qty in res.iteritems():
                 res[product_id]['immediately_usable_qty'] = \
-                    stock_qty['qty_available'] + stock_qty['outgoing_qty']
+                    stock_qty['qty_available'] - stock_qty['outgoing_qty']
```
This seems correct, but it scares me. Was it this wrong on the original 7.0 version?

gfcapalbo replied:
In 8.0 outgoing qty is calculated as an absolute quantity.
https://github.com/odoo/odoo/blob/8.0/addons/stock/product.py#L380-L381
https://github.com/odoo/odoo/blob/8.0/addons/stock/product.py#L112-L152

In 7.0 that wasn't the case amounts are subtracted, and returned as a negative value:
https://github.com/odoo/odoo/blob/7.0/addons/stock/product.py#L334-338

therefore our interpretation of immediately_usable_qty needed to be changed to keep uniformity in user experience between versions.

clonedagain replied:
In the newest v7 this was changed to available immediately = available - incomming.
I hope it's ok for you?

lepistone commented:
Another remark: the modification of `field_names` with the long commentary seems unnecessary: I gave a look to the original `_product_available` method in the `stock` module and it doesn't use `field_names` at all.

lepistone commented:
The first commit has been done with an incorrect author. That can complicate logs and attribution. Can you fix that with an interactive rebase and `git commit --amend --author` ? (`git filter-branch` should also work). Thanks!

StefanRijnhart commented on stock_reserve/view/product.xml:
```diff
@@ -1,25 +1,10 @@
-                </xpath>
-            </field>
-        </record>
-
```
As this stanza was deleted, does this mean that you can not inspect the number of reserved product variants?

StefanRijnhart replied:
In fact, maybe the changes to stock_reserve are unintended as it was already ported in revision c6d897dadbe208c774b9274e568b003c713b950c and the rest of this branch concerns another module?

gfcapalbo replied:
it's been made useless by the modified:
https://github.com/Therp/stock-logistics-warehouse/blob/stock_available_immediately2/stock_reserve/view/product.xml#L4-L16

that serves the same function.

StefanRijnhart replied:
Thanks for the clarification. Can you please move this change to a separate PR so that we can deal with both issues separately?

gfcapalbo commented:
Pull request closed, moved to #35, sorry for the inconvenience.

gfcapalbo commented:
nevermind , we will continue on #33, the original pull request.

clonedagain commented:
Why is there a change in stock_reserve?

StefanRijnhart commented:
Setting to Needs fixing as per @clonedagain's comment given the earlier discussion about changes in stock_reserve.

gfcapalbo commented:
Stock_reserve modifications have been reverted.

clonedagain commented:
@gfcapalbo I'll try to push an updated branch for you to merge, to match the changes committed yesterday in v7.
I'll to it this morning if I can.

clonedagain commented:
Easier said than done...
The cleanest way would have been to merge the latest v7 commit in your branch, but `7.0` has never been forward-ported to `8.0` yet.
I'll push a branch with just the matching patch. I'm sure we can do better but what?

clonedagain commented on stock_available_immediately/product.py:
```diff
@@ -20,42 +20,24 @@
     """
-    _inherit = 'product.product'
+    _inherit = 'product.template'
 
```
Why is it `product.template` now?

gfcapalbo replied:
Not strictly necessary, following the tendency of odoo to put most product.product fields in the template.

clonedagain commented:
@gfcapalbo, I've pushed https://github.com/Therp/stock-logistics-warehouse/pull/1 to help you integrate the latest changes from v7. Please review it and merge it into your branch.

lepistone committed:
make stock_reserve owner-aware

lepistone committed:
make stock_reserve_sale owner_aware

StefanRijnhart commented:
@clonedagain It is probably too late now, but I don't understand why this PR should include stock_available_mrp and stock_available_sale now all of a sudden. What is the reason? I would like to see PRs to deal with as little unrelated changes as possible.

StefanRijnhart commented:
I don't think that inheriting product.template instead of product.product is cutting it for this module's migration to 8.0. I am testing with a 'product' (i.e. product.template) with two 'variants' (i.e. product.product). There are no incoming or outgoing goods. I'm setting the stock of one variant to two items on stock, the other to three. This results in an immediately available quantity of 5 on both!

This confusion seems to be caused by counting products instead of variants. If you look at addons/stock/product.py, you see that fields like outgoing_qty are now defined both on product.template as on product.product, each with their own function. You need to do the same for these modules.

clonedagain commented:
@StefanRijnhart for consistency I've cherry-picked all the commits of the (admittedly massive) PR merged in v7, that's why *_mrp and *_sale are included.
I thought it would be difficult to bring in some modules now and others later, given that all impact the config file of the generic module "stock_available".

Maybe it would be simpler if we could just merge 7.0 into 8.0 once in a while?
If others are willing to help with this, I could contribute PRs to review the conflict resolutions.

StefanRijnhart commented:
@clonedagain thanks for clarifying. I guess I can still easily review them by diffing each of these modules with the 7.0 branch.

Forward merging would be really nice. Unfortunately, because of $#@!^ git's lack of support for tracking moved files and directories, merging into a repo that still keeps modules in **unported** will probably be too much work.

lepistone committed:
fix #42: reserve location should be outside WH

This has always been wrong on v8, but because of the very obscure
odoo/odoo#5797, this seemed to work every time we had single-step
reception.

Please note that this block of XML is noupdate, so for existing
installations you need to change the parent of the reservation location
to something outside your warehouses manually.

Incidentally, this makes the branch green independently of odoo/odoo#5797.

lepistone committed:
use context when creating the wizard

Because the context is used in the default methods now.

lepistone committed:
improve float assertions

gfcapalbo commented:
Pushed review after @StefanRijnhart  observation about product.template, it seems that now  that _immediately_usable_qty is computed correctly for product.template and product.product/variants.

StefanRijnhart commented:
Thanks. Tests are failing as indicated by Travis CI but these are in unrelated modules. 

However, there are still some leftovers from the merge-in of the 7.0 branch by @clonedagain: both stock_available_sale and stock_available_mrp appear in **unported** as well as in the repo root. The latter contain only the README.rsts. @clonedagain, is this on purpose maybe related to the recent **unported** discussion or do these files need to be moved?
There is another test that fails. It is in stock_available_sale.

clonedagain commented:
Not on purpose, I must have failed somewhere.
I'll try to make this right tomorrow if no-one can fix it before.

gfcapalbo commented:
@clonedagain I'll do it now.

lepistone committed:
bump and migrate

lepistone committed:
fix crash: excepted singleton

gfcapalbo commented:
ready for review.

lepistone committed:
find XML id by module, name

gurneyalex committed:
don't use travis container-based infrastructure

gurneyalex committed:
Merge pull request #43 from lepistone/fix-reservation-location

fix #42: reserve location should be outside WH

lepistone committed:
fix reservation_count computation

1. Make it float, because quantity is float.
2. Respect unit of measure (instead of adding apples and pears).
3. Do that for both product and template.
4. Test the above.

lepistone committed:
refactor: delegate to variants, use self.id

lepistone committed:
refactor tests, improve failures

yvaucher committed:
Merge pull request #44 from lepistone/fix-expected-singleton

fix crash: excepted singleton

Ehtaga committed:
Add partner_area_data and partner_area_management

This module was written to enter a location's area based on different units of measure
The information is dependant on the different locations entered

Ehtaga committed:
Rename modules with stock_location instead of partner

jgrandguillaume committed:
Merge pull request #41 from lepistone/reserve-owner

make stock_reserve and stock_reserve_sale owner_aware

Ehtaga committed:
Add units of measure conversion

gurneyalex committed:
Update travis configuration

Follow OCA/maintainer-quality-tools#187 and use addons/apt/package directive to
install our build dependencies. This allows using container based Travis builds
(which should be faster).

Commit:
Add README.rst and remove description from __openerp__.py

Commit:
Add README.rst and remove description from __openerp__.py

Commit:
Add README.rst and remove description from __openerp__.py

pedrobaeza committed:
Merge pull request #53 from Doscaal/8.0-stock-reserve_readme

Add README.rst and remove description from __openerp__.py

Commit:
Merge pull request #54 from Doscaal/8.0-stock-location-ownership_readme

[IMP] Add README.rst and remove description from __openerp__.py

yvaucher committed:
Merge pull request #55 from Doscaal/8.0-stock-reserve-sale_readme

Add README.rst and remove description from __openerp__.py

pedrobaeza committed:
Merge pull request #46 from savoirfairelinux/partner_area_management

Add partner_area_data and partner_area_management

gurneyalex commented:
:+1: although a small automated test would be very welcome.

gurneyalex commented:
@clonedagain could  you rebase the branch to trigger a runbot build?

clonedagain commented:
@gurneyalex sorry I can't, I don't own this branch. @gfcapalbo can you do it please?

gurneyalex committed:
Merge Stock available immediately2

gurneyalex commented:
@clonedagain and @gfcapalbo I started #57 to finalize this PR -> runbot will be available there.

StefanRijnhart commented:
@gurneyalex we are working today on tests for this module. Can you give us another day?

gfcapalbo committed:
[ADD] tests

gurneyalex commented:
@StefanRijnhart @gfcapalbo sure, I'm always happy to not do work. And I've fixed a bug which was preventing the runbot from starting builds this morning, so we should be shortly green.

Thanks for your time.

gurneyalex commented:
:+1:

StefanRijnhart commented on stock_available_immediately/tests/test_stock_available_immediately.py:
```diff
@@ -0,0 +1,120 @@
+
+    def test01_stock_levels(self):
+        """checking that immediately_usable_qty actually reflects \
+the variatins in stock, both on product and template"""
```
Typo in 'variations'

gfcapalbo committed:
[Fix] Typo

gfcapalbo commented:
Fixed typo.

StefanRijnhart commented:
:+1:

gfcapalbo committed:
[ADD} location calculations

gfcapalbo committed:
[FIX] typo

gfcapalbo committed:
[DEL] removed references to product.py, not needed anymore, the quant
calculations done in product template are fine.

gfcapalbo committed:
[DEL] remove old tests, rewriting

Commit:
[FIX] searching on products, leaving full calculations
not using internal qty_available that seems not to
take in consideration reserved quants.

Commit:
[ADD] Tests

gurneyalex commented on stock_available/product.py:
```diff
@@ -0,0 +1,66 @@
+#-*- coding: utf-8 -*-
```
adding a space after `#` will make flake8 happy and travis :green_heart: :

```
./stock_available/product.py:1:1: E265 block comment should start with '# '
```
